### ISSUE TITLE
matching against an associated constant gives type error

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#57280]
Ever since #55937, the following program ([playground][]) gives an unexpected error from NLL:

[playground]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c864079c6a555e7960818b2fa82690cb

```rust
#![feature(nll)]

trait Foo {
    const Blah: &'static str;
}

struct Placeholder;

impl Foo for Placeholder {
    const Blah: &'static str = "hi";
}

fn foo(x: &str) {
    match x {
        <Placeholder as Foo>::Blah => { }
        _ => { }
    }
}
```

First reported by @Xanewok [in this comment](https://github.com/rust-lang/rust/pull/55937#issuecomment-450899521), because this broke the RLS.

cc @davidtwco

**Comment by nikomatsakis** [comment]
I dug into this a bit -- see the [zulip log] for full details. In short, the problem is:

- In https://github.com/rust-lang/rust/commit/4be7214d30b121762c8675c3e732083da262b136, we added a check that the type `T` of the value being matched must be a subtype of the type `C` of the constant
- The lexical type checker has the reverse: `C <: T` ([link](https://github.com/rust-lang/rust/blob/ec194646fef1a467073ad74b8b68f6f202cfce97/src/librustc_typeck/check/_match.rs#L153-L165))
- Really neither is quite right, but we can match the lexical type checker by including a notion of "variance" on the pattern type annotations

Probably what we *really* want is to require that `T: PartialEq<C>`.

[zulip lop]: https://rust-lang.zulipchat.com/#narrow/stream/122657-wg-nll/subject/.2354943.20user.20type.20annot.20in.20unreachable.20code/near/154203643
